staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,570 @@
+	}
+	metrics.ObserveQueueLength(qs.config.Name, len(queue.Requests))
+	return req
+}
+
+// removeTimedOutRequestsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *queueSet) removeTimedOutRequestsFromQueue(queue *fq.Queue) {
+	timeoutIdx := -1
+	now := qs.clock.Now()
+	reqs := queue.Requests
+	// reqs are sorted oldest -> newest
+	// can short circuit loop (break) if oldest requests are not timing out
+	// as newer requests also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.config.RequestWaitLimit)
+	for i, req := range reqs {
+		if waitLimit.After(req.RealEnqueueTime) {
+			qs.counter.Add(1)
+			req.DequeueChannel <- false
+			close(req.DequeueChannel)
+			// get index for timed out requests
+			timeoutIdx = i
+		} else {
+			break
+		}
+	}
+	// remove timed out requests from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout req
+		removeIdx := timeoutIdx + 1
+		// remove all the timeout requests
+		queue.Requests = reqs[removeIdx:]
+		// decrement the # of requestsEnqueued
+		qs.numRequestsEnqueued -= removeIdx
+	}
+}
+
+// getRequestsExecutingLocked gets the # of requests which are "executing":
+// this is the# of requests/requests which have been dequeued but have not had
+// finished (via the FinishRequest method invoked after service)
+func (qs *queueSet) getRequestsExecutingLocked() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// chooseQueueIndexLocked uses shuffle sharding to select a queue index
+// using the given hashValue and the shuffle sharding parameters of the queueSet.
+func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64) int {
+	bestQueueIdx := -1
+	bestQueueLen := int(math.MaxInt32)
+	// DesiredNum is used here instead of numQueues to omit quiescing queues
+	qs.dealer.Deal(hashValue, func(queueIdx int) {
+		thisLen := len(qs.queues[queueIdx].Requests)
+		if thisLen < bestQueueLen {
+			bestQueueIdx, bestQueueLen = queueIdx, thisLen
+		}
+	})
+	return bestQueueIdx
+}
+
+// updateQueueVirtualStartTime updates the virtual start time for a queue
+// this is done when a new request is enqueued.  For more info see:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#dispatching
+func (qs *queueSet) updateQueueVirtualStartTime(request *fq.Request, queue *fq.Queue) {
+	// When a request arrives to an empty queue with no requests executing:
+	// len(queue.Requests) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.Requests) == 1 && queue.RequestsExecuting == 0 {
+		// the queue’s virtual start time is set to the virtual time.
+		queue.VirtualStart = qs.virtualTime
+	}
+}
+
+// enqueues a request into an queueSet
+func (qs *queueSet) enqueue(request *fq.Request) {
+	queue := request.Queue
+	queue.Enqueue(request)
+	qs.updateQueueVirtualStartTime(request, queue)
```

Comment:
I would have expected this update to happen when we start *executing* the request, not when we enqueue it?

Reply:
It looks like you are saying a pointer to the KEP is not good enough, we have to copy the reasoning from there into the code.

Reply:
I expect this can be explained in one sentence and that sentence would be super enlightening. Worst case, the sentence is "there's no possible short explanation of this, see the KEP"; that would tell me that there's something major missing in my mental model of how this code works.

But I expect that a useful sentence like "The virtual clock on a queue starts when the first request is queued (and e.g. not when the request begins to execute) because ____." can be written.